[Inductor] Enable VecMask store

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123710
* #123512

**Summary**
Enable the vectorization of store with `bool` dtype. 

**Test Plan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel 
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

leslie-fang-intel committed:
[Inductor] Enable VecMask store

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable VecMask store"


**Summary**
Enable the vectorization of store with `bool` dtype. 

**Test Plan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel 
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable VecMask store"


**Summary**
Enable the vectorization of store with `bool` dtype. 

**Test Plan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel 
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -2881,6 +2881,7 @@ def __init__(self, args, num_threads, tiling_factor, tiling_idx=-1):
             torch.int8,
             torch.int32,
             torch.int64,
+            torch.bool,
```
with this change, we can use one list `supported_dtypes` for all such checks since the lists are all the same now. Can we define one list and used by all the checks in the code?

leslie-fang-intel replied:
Thanks for the comment. Merge into one list of `supported_dtypes`.

jgong5 commented on test/inductor/test_cpu_repro.py:
```diff
@@ -1825,6 +1825,8 @@ def run_compile_decomposed_fake_quant(
         self.assertEqual(input_grad_aten_eager, input_grad)
         self.assertEqual(input_grad_decomp_eager, input_grad)
         self.assertEqual(input_grad[1, 2, 3, 4], torch.tensor(0.0))
+        # For forward and backward kernel
+        check_metrics_vec_kernel_count(2)
```
Is this a related change? Do we have UT on vectorized bool?

leslie-fang-intel replied:
Yean. We have a UT `python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel `. Previously, the forward graph has been disabled of vectorization due to the store of bool. And after this PR, we can vectorize the forward graph now.

jgong5 commented on aten/src/ATen/cpu/vec/vec_mask.h:
```diff
@@ -134,6 +134,14 @@ class VecMask {
+  void store(bool* b, int count) {
+    TORCH_CHECK(count == Vectorized<T>::size() * N, "Expect same number of elements in VecMask::store.");
+    TORCH_CHECK(count <= Vectorized<bool>::size(), "Expect the number is less or equal to Vectorized<bool>.");
+    Vectorized<bool> res = this->to<bool, 1>();
```
```c++
constexpr int L = (VectorizedN<T, N>::size() + Vectorized<bool>::size() - 1)/ Vectorized<bool>::size();
auto res = this->to<bool, L>();
res.store(b, count);
```

leslie-fang-intel replied:
Thanks for the comments.

jgong5 commented on aten/src/ATen/cpu/vec/vec_mask.h:
```diff
@@ -134,6 +134,14 @@ class VecMask {
   }
 
+  void store(bool* b, int count) {
+    TORCH_CHECK(count == Vectorized<T>::size() * N, "Expect same number of elements in VecMask::store.");
+    TORCH_CHECK(count <= Vectorized<bool>::size(), "Expect the number is less or equal to Vectorized<bool>.");
```
No need to explicit assert here? It should be caller's responsibility to guarantee the count is valid?

leslie-fang-intel replied:
Sure and removed.

jgong5 commented on aten/src/ATen/cpu/vec/vec_mask.h:
```diff
@@ -134,6 +134,14 @@ class VecMask {
     return VectorizedN<T, N>(VectorizedN<T, N>::loadu(mask));
   }
 
+  void store(bool* b, int count) {
```
```suggestion
  void store_mask(bool* b, int count) {
```

leslie-fang-intel replied:
There are several places in cpp.py using `store` instead of `store_mask`, such as:

- https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/torch/_inductor/codegen/cpp.py#L2449-L2453
- https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/torch/_inductor/codegen/cpp.py#L2343
- https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/torch/_inductor/codegen/cpp.py#L2815

I afraid we need to check the dtype in all the places and change `store` to `store_mask`?

jgong5 replied:
oh, that's good point. I was just concerned that the name might be confusing with "masked store" semantics. Perhaps, we'd better rename `storeu` with `masked_storeu`..

leslie-fang-intel committed:
Update on "[Inductor] Enable VecMask store"


**Summary**
Enable the vectorization of store with `bool` dtype. 

**Test Plan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel 
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -2902,7 +2892,7 @@ def load(self, name: str, index: sympy.Expr):
                 return var
 
-            if load_dtype not in self.load_supported_dtypes and (
+            if load_dtype not in self.supported_dtypes and (
```
Also check things in `constant` and `index_expr`?

leslie-fang-intel replied:
Unify the list in `constant`, `to_dtype`. I didn't see the dtype list in `index_expr`.

jgong5 replied:
It is a typo.

leslie-fang-intel committed:
Update on "[Inductor] Enable VecMask store"


**Summary**
Enable the vectorization of store with `bool` dtype. 

**Test Plan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel 
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable VecMask store"


**Summary**
Enable the vectorization of store with `bool` dtype. 

**Test Plan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel 
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jgong5 @CaoE, the `blendv` for `VecMask` have been added. Please kindly help to take a look again.

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
Hi @lezcano, could you help to take a look of this PR?

leslie-fang-intel commented:
Hi @peterbell10, Could you help to take a look of this PR?

lezcano reviewed (approved):
Sorry, I was on PTO after today. I'm trying to catch-up on my PRs

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
> Sorry, I was on PTO after today. I'm trying to catch-up on my PRs

Thanks for looking into this PR and enjoy the time off.
